File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  defp yelling?(input) do
    letters =
      input
      |> String.graphemes()
      |> Enum.filter(&String.match?(&1, ~r/^\p{L}$/u))

    letters != [] and Enum.all?(letters, &(&1 == String.upcase(&1)))
  end

  defp question?(input) do
    input |> String.trim() |> String.ends_with?("?")
  end

  defp silence?(input) do
    String.trim(input) == ""
  end

  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      question?(input) and yelling?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      yelling?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end
end

File: solutions/clojure/bob/1/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn response-for [s]
  (let [trimmed (str/trim s)
        is-question (str/ends-with? trimmed "?")
        is-silence (str/blank? trimmed)
        has-letters (some Character/isLetter trimmed)
        is-yelling (and has-letters (= (str/upper-case trimmed) trimmed))]
    (cond
      is-silence "Fine. Be that way!"
      (and is-yelling is-question) "Calm down, I know what I'm doing!"
      is-yelling "Whoa, chill out!"
      is-question "Sure."
      :else "Whatever.")))
